common/src/main/scala/com/gu/media/upload/UploadLifecycle.scala

```diff
@@ -0,0 +1,88 @@
+package com.gu.media.upload
+
+import java.util.UUID
+
+import com.amazonaws.services.s3.model.{CompleteMultipartUploadRequest, CopyPartRequest, InitiateMultipartUploadRequest, PartETag}
+import com.gu.media.aws._
+import com.gu.media.logging.Logging
+import scala.collection.JavaConverters._
+
+class UploadLifecycle(aws: AwsAccess with S3Access with DynamoAccess with KinesisAccess with UploadAccess) extends Logging {
+  private val credentials = new CredentialsGenerator(aws)
+
+  def create(metadata: UploadMetadata, size: Long): Upload = {
+    val id = UUID.randomUUID().toString
+    val parts = chunk(id, size)
+
+    Upload(id, parts, metadata, error = None)
+  }
+
+  def credentialsForPart(upload: Upload, part: UploadPart): UploadCredentials = {
+    credentials.forKey(upload.id, part.key)
+  }
+
+  def partComplete(upload: Upload, part: UploadPart): Upload = {
+    val complete = upload.withPart(part.key)(_.copy(uploadedToS3 = true))
+
+    // Put on Kinesis to be picked up by the UploadActionsLambda
+    val action = UploadPartToYouTube(upload.id, part.key)
+    aws.sendOnKinesis(aws.uploadActionsStreamName, upload.id, action)
+
+    if(complete.parts.forall(_.uploadedToS3)) {
+      val fullKey = createFullObject(complete).toString
+      val action = DeleteParts(upload.id, upload.parts.map(_.key))
+
+      aws.sendOnKinesis(aws.uploadActionsStreamName, upload.id, action)
+      sendToPluto(fullKey, upload.metadata)
+    }
+
+    complete
+  }
+
+  private def chunk(uploadId: String, size: Long): List[UploadPart] = {
+    val boundaries = Upload.calculateChunks(size)
+
+    boundaries.zipWithIndex.map { case ((start, end), id) =>
+      UploadPart(UploadPartKey(aws.userUploadFolder, uploadId, id).toString, start, end)
+    }
+  }
+
+  private def createFullObject(upload: Upload): UploadFullKey = {
```

Comment:
I renamed it to `createCompleteObject` and the corresponding URI class to `CompleteUploadKey`